ztp/source-crs-builder/MachineConfigChronyDynamicMaster/ptp-sync-check

```diff
@@ -0,0 +1,157 @@
+#!/bin/bash
+# Evaluate PTP synchronization state and create or delete 
+# the chronyd.service condition file accordingly
+
+# The parametes below are tuned using chronyd-restart.service drop-in
+# Maximum PTP offset to be considered in-sync (ns)
+MAX_OFFSET=${MAXIMUM_WAIT_TIME:-100000}
+# PTP offset hysteresis threshold (ns)
+OFFSET_THRESHOLD=${OFFSET_THRESHOLD:-20000}
+# Maximum valid measurement age (s)
+SYNC_AGE=${SYNC_AGE:-600}
+
+# File for starting chronyd conditionally
+SYNC_MARKER_PATH="/var/run/ptp/insync"
+
+# Auxiliary logging function
+function log_debug(){
+  local LOGGER_TAG="ptp-chronyd"
+  echo "$1"
+  logger -t $LOGGER_TAG -p "syslog.debug" "$1"
+}
+
+# Compare floating point numbers
+function greater() {
+   awk -v n1="$1" -v n2="$2" 'BEGIN {printf (n1>n2?"true":"false")"\n", n1, n2}'
+}
+
+# Return absolute (for master offset)
+function abs() { 
+    echo ${1##-}
+}
+
+# Check variable is defined
+function var_defined() {
+  VAR="$1"
+  if [[ ${#VAR} -eq 0 ]]; then 
+    log_debug "Early exit at line $(caller). PTP might not be running"    
+    return 1
+  fi
+  return 0
+}
+
+# Get the last line containing masteroffset from the latest available log
+# ARGUMENTS:
+#   none
+# GLOBAL:
+#   none
+# OUTPUT:
+#   LAST_STATUS - Last line of the log containing offset and timestamp
+#     Something like "2021-10-18T16:47:06.022983665+00:00 stdout F ptp4l[8180.150]: [ens4f0] 
+#     master offset 3 s2 freq -16756169 path delay 1481"
+# RETURN:
+#   0 if all sanity checks passed, 1 otherwise
+get_last_ptp_status() {
+  LOG_SUBDIR=$(ls -tr /var/log/pods/ |grep openshift-ptp_linuxptp-daemon |tail -1)
```

Comment:
So this gives us the newest directory.

Would it be better to query `crictl pods` to look for a matching pod so we only find the one that's currently running? Or is the newest log dir good enough?

Reply:
`crictl pods` gives me the pod, but not the directory it's logging to:
```bash
[root@cnfdt16 core]# ls /var/log/pods/ | grep linuxptp
openshift-ptp_linuxptp-daemon-lthj2_22cce38c-a99d-4818-ab63-3dbe6558e03c
```
versus
```bash
[root@cnfdt16 core]# crictl pods |grep ptp
3231b0544dec3       18 hours ago        Ready               ptp-operator-8c65cd8f6-p7wjx                                 openshift-ptp                                      0                   (default)
946589188a99b       18 hours ago        Ready               linuxptp-daemon-lthj2                                        openshift-ptp                                      0                   (default)
```

Reply:
We can use crictl  to get the log path!
1. Get the linuxptp daemon container id:
```bash
# crictl ps --name linuxptp-daemon-container --latest --state Running --quiet
6947569dd2f1d247ca8e6018eae908bb8087fe77cea01643ae6763880fa34054
```
2. Get the container log path:
```bash
 # crictl inspect -o json 6947569dd2f1d247ca8e6018eae908bb8087fe77cea01643ae6763880fa34054 |jq '.status.logPath'
"/var/log/pods/openshift-ptp_linuxptp-daemon-lthj2_22cce38c-a99d-4818-ab63-3dbe6558e03c/linuxptp-daemon-container/16.log"
```
I was also trying to get away by using `crictl logs` command, but I had an issue with piping the output to `tail`. I noticed that the authors added these to the `crictl log` command options, but I wanted to pipe to tail twice: first to take a reasonable size buffer, then filter it with grep, and then take the latest line. With built-in --tail it doesn't work... Do you have an idea how can I work around this?

Reply:
We can also get away exclusively with crictl:
```bash
# crictl logs --since 1m --timestamps 6947569dd2f1d247ca8e6018eae908bb8087fe77cea01643ae6763880fa34054 |grep -m 1 "master offset"
2021-10-19T15:30:41.031729724Z ptp4l[5353.967]: [ens4f0] master offset        -16 s2 freq   +9455 path delay      1460

```
I'm still checking this